### ISSUE TITLE
Upgrade to 2.0.x causes error

### ISSUE THREAD

**Issue by typeoneerror** [mharris717/ember-cli-pagination#146]
Not sure if there is something to be done in pagination, but at the moment it looks incompatible with Ember 2.0.x because of an issue with query params:

https://github.com/emberjs/ember.js/issues/11592

If I use the *Binding methods I run into this error.

**Comment by DylanLukes** [comment]
+1, will update with more as I investigate this.

**Comment by broerse** [comment]
@mharris717 like to update to 2.0 soon. @DylanLukes if you switch to 2.0 are there errors in `ember test` also?

**Comment by DylanLukes** [comment]
The issues I'm having seem to be related to query parameters and binding, as discussed in the linked issue. This isn't due to a *bug* so much as depending on behavior no longer supported by Ember. And to my understanding, it wasn't officially supported in the first place, it just happened to work.

Here's my understanding of what's happening:

1) Route returns model, `router.js` calls into `Controller#setup` (excerpt below).
2) Since there is a transition (from `''` to `<your-model>.index`) it also updates the query parameters.
3) For some reason, the `cache` is found, so it iterates over the query params stored in it and sets them on the controller.
4) When it attempts to set query parameters on the controller it triggers `Controller#_qpChanged`.
5) However, `_qpChanged` depends on `_qpDelegate` being set, and that doesn't happen until the end of the first excerpt.

Hope this helps. I'm still actively investigating currently.

Except 1:
```js
var states = queryParams.states;
      if (transition) {
        // Update the model dep values used to calculate cache keys.
        _emberRoutingUtils.stashParamNames(this.router, transition.state.handlerInfos);

        var params = transition.params;
        var allParams = queryParams.propertyNames;
        var cache = this._bucketCache;

        allParams.forEach(function (prop) {
          var aQp = queryParams.map[prop];

          aQp.values = params;
          var cacheKey = _emberRoutingUtils.calculateCacheKey(aQp.prefix, aQp.parts, aQp.values);

          if (cache) {
            var value = cache.lookup(cacheKey, prop, aQp.undecoratedDefaultValue);
            _emberMetalProperty_set.set(controller, prop, value);
          }
        });
      }

      controller._qpDelegate = states.allowOverrides;
```

Except 2:
```js
_qpChanged: function (controller, _prop) {
      var prop = _prop.substr(0, _prop.length - 3);

      var delegate = controller._qpDelegate;
      var value = _emberMetalProperty_get.get(controller, prop);
      delegate(prop, value);
    },
```

**Comment by DylanLukes** [comment]
Good news everyone.

I have a fix. It's a little involved but can be integrated pretty easily. I'm nailing down the details. The changes are roughly:

1) remove pageBinding/perPageBinding from controller. You still need the totalPage binding and that's totally kosher.
2) do not update PagedRemoteArrays page/perPage. Instead, add actions to your router that modify query parameters and force a reload of the model ('refreshModel: true' on query params on route)

To summarize, changes in page/perPage should propagate from the router down via model refreshes. Changes in the page/perPage parameters should not be the concern of the Controller, but the Router, which should update them based on actions.

**Comment by jcope2013** [comment]
:+1:

**Comment by DylanLukes** [comment]
Currently rewriting a new `PaginationComponent`. The controller/route mixins aren't strictly necessary...

**Comment by mharris717** [comment]
Anyone have interest in pairing with me on this? I apologize for being a shit maintainer. 

@jcope2013 @DylanLukes @broerse

**Comment by DylanLukes** [comment]
Unfortunately, I can't currently take up maintaining anything.

1) I'm no Ember expert.
2) I'm currently an undergrad, applying to graduate schools.

It looks like for the most part it should be possible to reimplement this library very minimally, though. I hope my samples will demonstrate the right direction.

**Comment by mharris717** [comment]
@DylanLukes just looking for somebody to work with me on a single pairing session. 

Thanks for all your work already.

**Comment by DylanLukes** [comment]
Ah okay, I don't actually know what a pairing session is :). I haven't been formally trained in software engineering. I'm a programmer, but I'm not a software developer. I just write web apps and infrastructure for my part-time job.

**Comment by mharris717** [comment]
@DylanLukes based on your comments here you seem to know what you're doing :)

Was looking for somebody to screenshare with and work through getting this upgraded to 2.0 together. Apparently I can't muster up the activation energy on my own.

**Comment by DylanLukes** [comment]
I may be able to later on, but probably not any time soon. Graduate applications have consumed my life.

From the looks of it though, all that's really necessary is a nice RouteMixin and a PaginationComponent. In general, I think the PagedArray classes can go, that state management should probably be kept in the route. 

The data flow should be:

actions (nextPage, prevPage)
=> linkTo/transitionTo
=> URL Params 
=> refreshModel
=> Controller#model
=> Controller#setupController: setAttributes on PaginationComponent (and display Component or setModel on a View for legacy)

Basically, propagate data downwards only. The controller just needs a dumb model with the current page. The other state (currentPage, etc) can be passed down in an opt-in manner via separate attributes on the component/view displaying the paginated content.

tl;dr: Model loading is a Router concern, and Pagination is a dependency of Model loading, so it's also a Router concern.

**Comment by mharris717** [comment]
The PagedArray classes were my attempt to keep some OOP sanity, instead of spreading the logic all around the route. It might be too much trouble, I don't know.

**Comment by DylanLukes** [comment]
I see that, and for the most part that works great. It's just Ember is moving away from fat models :\.

What I'm thinking for a rewrite (I may just do one myself) is:

 - Router fetches data using current pagination query parameters. 
 - The data ([myModel]) is split from the metadata (currentPage, totalPages, ...)
 - The actual entities are passed as a standard Ember.Array to a component, with the metadata passed separately if desired.

(Side note: I'd probably prefer `limit`/`offset` over `perPage`/`page`/... but I use Django so that's unsurprising ;))

Here's a (non-functioning) quick sketch of how I see this working:

controllers/widgets.js
```js
import Ember from "ember";

export default Ember.Controller.extend({
  page: 1,
  perPage: 5,

  queryParams: ["page", "perPage"]
});
```

routes/widgets.js
```js
import Ember from "ember";

export default Ember.Route.extend({
  queryParams: {
    page: {refreshModel: true},
    perPage: {refreshModel: true},
  },
  
  model: function(params) {
    return Ember.RSVP.hash({
      content: this.store.find("widget", {
        page: params.page || 1,
        per_page: params.perPage || 10
      }
    });
  },

  setupController: function(controller, models) {
    controller.setProperties(models);
  },

  actions: {
    nextPage: function() {
      this.transitionTo('widgets', {page: this.modelFor('widget').get('meta.page') + 1});
    },
    prevPage: function() {
      this.transitionTo('widgets', {page: this.modelFor('widget').get('meta.page') - 1});
    },

    createWidget: function() {
      var newWidget = ...;
      // ...
      this.transitionTo('post', newWidget.save());
    },

    saveWidget: function() {
      this.modelFor('widget').save();
    },

    deleteWidget: function() {
      this.modelFor('widget').destroyRecord().then(function() {
        this.transitionTo('widgets');
      }.bind(this));
    }
  }
});
```

templates/widgets.hbs
```hbs
{{#widgets-master widgets=content page=content.meta.page ... }}
{{#pagination-thingy page=content.meta.page ...}}
```

We can override `extractMeta` on the Route to replace the old paramMapping and support different parameter schemes. :smile:

**Comment by DylanLukes** [comment]
Note this is agnostic to the pagination style. For remote pagination you'd do it like that, and for local pagination you do the same thing but omit the parameters in find() and instead define a "pagedContent" property on the controller.

Remote pagination happens in the router , local happens in the controller via computed properties.

Metadata management, instead of being placed on the content itself via a PagedArray, can just use Embers built in metadata management.

**Comment by jcope2013** [comment]
@DylanLukes why is the
 ``` 
setupController: function(controller, models) {
    controller.setProperties(models);
  },
``` 
needed? or what is the intention of that? I am sort of confused there

**Comment by DylanLukes** [comment]
Oh, that's just a general practice. The assumption is that one route may need to load multiple models or sets of models to render disparate components.

You could do for instance:

```js
  model: function(params) {
    return Ember.RSVP.hash({
      content: this.store.findAll('widget'), // the 'primary' model, aliased to 'model'
      gadgets: this.store.findAll('gadget'),
    });
  },

  ...

  setupController: function(controller, models) {
    controller.setProperties(models);
  },

  // now we just have to be explicit and use this.controllerFor and this.modelFor throughout
  // get('controller') or get('model') will return the ones for 'content'.
```

, and then (using a barebones controller as before) we have a template like:

```
{{#widget-thing widgets=content}}{{outlet}}{{/widget-thing}}
{{#gadget-aux-thing gadgets=gadgets}}
```

But `gadgets` is in scope here for auxiliary components, and we can access belongsTo relations inside the components without implicit scope violations. It's just, I think, more semantically correct and cleaner.

**Comment by DylanLukes** [comment]
Working on something for my own project. I'll factor it out into https://github.com/DylanLukes/ember-cli-paged . I don't intend on maintaining it, as pagination will be hitting ember-data. See:

https://github.com/emberjs/data/issues/2905
https://github.com/emberjs/data/issues/3048

**Comment by DylanLukes** [comment]
Here's some snippets for those dealing with issues. I've managed to remove the need for server side pagination, at least for now. It might not be perfect yet, and notably the template is incomplete and doesn't provide gaps or sliding windows. How you render your page numbers is up to you.

This code handles pagination for a full set of available objects. It handles orphans also. It's a shameless port of Django's pagination. You can easily modify this to include gaps and modify the rendering. I've also implemented `Ember.Array` on these components as well.

`/pods/better-page-numbers/component.js`
```js
import Ember from 'ember';

var Page = Ember.Object.extend(Ember.Array, {
    objects: [],
    pageNumber: null,
    paginator: null,

    hasNext: Ember.computed.lt('pageNumber', 'paginator.pageCount'),

    hasPrev: Ember.computed.gt('pageNumber', 1),

    hasOther: Ember.computed.or('hasPrev', 'hasNext'),

    nextPageNumber: function () {
        return this.get('paginator').validatePageNumber(this.get('number') + 1);
    }.property('paginator', 'pageNumber'),

    prevPageNumber: function () {
        return this.get('paginator').validatePageNumber(this.get('number') - 1);
    }.property('paginator', 'pageNumber'),

    startIndex: function () {
        if (this.get('paginator.count') === 0) {
            return 0;
        }
        return (this.get('paginator.perPage') * (this.get('pageNumber') - 1)) + 1;
    }.property('paginator.objectCount', 'paginator.perPage', 'pageNumber'),

    endIndex: function () {
        if (this.get('pageNumber') === this.get('paginator.pageCount')) {
            return this.get('paginator.objectCount');
        }
        return this.get('pageNumber') + this.get('paginator.perPage');
    }.property(''),

    // Page implements Ember.Array for objects
    length: Ember.computed.alias('objects.length'),
    objectAt: function (idx) {
        return this.get('objects').objectAt(idx);
    }
});

var SimplePaginator = Ember.Object.extend(Ember.Array, {
    objects: Ember.computed.oneWay('parent.objects'), // the set of all paginated objects
    perPage: Ember.computed.oneWay('parent.perPage'),
    orphans: Ember.computed.oneWay('parent.orphans'),  // orphans: the minimum number of items allowed on the last page.
    allowEmptyFirstPage: Ember.computed.oneWay('parent.allowEmptyFirstPage'),

    objectCount: Ember.computed.alias('objects.length'),

    validatePageNumber: function (number) {
        if (number % 1 !== 0) {
            throw new Ember.Error('That page number is not an integer');
        }

        if (number < 1) {
            throw new Ember.Error('That page number is less than 1');
        }

        if (number > this.get('pageCount')) {
            if (number !== 1 || !this.get('allowEmptyFirstPage')) {
                throw new Ember.Error('That page contains no results.');
            }
        }

        return number;
    },

    page: function (number) {
        var beginIndex = (number - 1) * this.get('perPage'),
            endIndex = beginIndex + this.get('perPage');

        if (beginIndex + this.get('orphans') >= this.get('objectCount')) {
            endIndex = this.get('objectCount');
        }

        return Page.create({
            pageNumber: this.validatePageNumber(number),
            objects: this.get('objects').slice(beginIndex, endIndex)
        })
    },

    pageCount: function () {
        if (this.get('objectCount') === 0 && !this.get('allowEmptyFirstPage')) {
            return 0;
        }
        else {
            let hits = Math.max(1, this.get('objectCount') - this.get('orphans'));
            return Math.ceil(hits / this.get('perPage'));
        }
    }.property('perPage', 'objectCount', 'allowEmptyFirstPage', 'orphans'),

    // SimplePaginator implements Ember.Array (by PAGES, not objects)
    length: Ember.computed.alias('pageCount'),
    objectAt: function (pageNumber) {
        return this.page(pageNumber);
    }
});

export default Ember.Component.extend({
    // Content to page, expected to be ember-cli-pagination/PagedArray
    pagedContent: null,
    perPage: 10,
    orphans: 0,
    objects: [],
    allowEmptyFirstPage: true,

    paginator: function() {
        return SimplePaginator.create({parent: this})
    }.property(),

    actions: {
        // These actions should propagate up to the router for when more data need to be loaded.

        next: function () {

        },
        prev: function () {

        },
        jump: function (pageNumber) {

        }
    }
});
```

`pods/better-page-numbers/template.hbs`
```hbs
<div class="pagination-centered">
    <ul class="pagination">
        <li class="arrow prev {{if paginator.hasPrev 'enabled-arrow' 'disabled'}} ">
            <a href="#">&laquo;</a>
        </li>

        {{!-- fill in here --}}

        <li class="arrow next {{if paginator.hasNext 'enabled-arrow' 'disabled'}} ">
            <a href="#">&raquo;</a>
        </li>
    </ul>
</div>
```

Hope this helps someone out.

**Comment by jcope2013** [comment]
i fixed the binding issue by removing the page/perPage bindings and instead adding a action that is already internally triggered by the page-numbers component when you change a page that updates the properties

```
export default Ember.Controller.extend({
  // setup our query params
  queryParams: ["page", "perPage"],

  totalPagesBinding: "content.totalPages",

  // set default values, can cause problems if left out
  // if value matches default, it won't display in the URL
  page: 1,
  perPage: 10,

  actions: {
     pageClicked(page, perPage) {
         this.setProperties({ page: page, perPage: perPage });
      }
   }
});

{{page-numbers content=pagedContent action=(action 'pageClicked')}}
```

**Comment by ianwalter** [comment]
@jcope2013's fix worked for me although I don't think `perPage` should be in the `pageClicked` function. Thanks for everyones work on this.

**Comment by thejchap** [comment]
@jcope2013's fix worked for me as well

```
DEBUG: Ember             : 2.2.0
DEBUG: Ember Data        : 2.3.1
```

**Comment by yanatan16** [comment]
:+1:  @jcope2013's fix worked for me. 
```
DEBUG: Ember             : 2.3.0
ember.debug.js:6271 DEBUG: Ember Data        : 2.3.3
```

**Comment by esbanarango** [comment]
@jcope2013 Thanks! forked for me as well.

**Comment by jme783** [comment]
+1 @jcope2013 thank you!

**Comment by lbblock2** [comment]
Any solutions for the remote paginated API?

**Comment by remino** [comment]
> Any solutions for the remote paginated API?

Also doesn't work for me. Getting this error message in my console:

```
ember.debug.js:28556 Error while processing route: [redacted controller name] delegate is not a function TypeError: delegate is not a function
```

**Comment by sergejgasjuk** [comment]
So, has anyone found a proper fix for this?

**Comment by lancedikson** [comment]
@jcope2013's solution works for remote paginated api as well

**Comment by typeoneerror** [comment]
@jcope2013's solution works but doesn't update the param in the URL for me. If I try to do what the documentation shows with `page: alias('content.page')`, I get 

> Property set failed: object in path "content" could not be found or was destroyed. Error: Property set failed: object in path "content" could not be found or was destroyed.

There seems to be a lot of conflicting setup information in docs, tests and issues. Does anyone have a definitive guide to making all the pieces of this addon work?

**Comment by typeoneerror** [comment]
Of note: https://github.com/emberjs/ember.js/issues/10608#issuecomment-78298761

QP's cannot be CP's, so the example in the documentation will not work. It looks like if I have 

```
queryParams: {
  page: { refreshModel: true }
}
```

this is the only way to get the page URL param to change. Is this what everyone is doing?

**Comment by typeoneerror** [comment]
Extending @jcope2013's solution, here's what's currently working for me:

```
setPage(page) {
  Ember.Logger.log('setPage', page);
  this.setProperties({
    'model.page': page,
    page: page
  })
}
```

Setting model.page triggers the fetch without reloading the model. Setting `page` triggers the URL param to change. I'm not able to just set one of the two with a computed property in any way.

**Comment by sdhull** [comment]
A few things here.

First, this addon as documented in the Readme will not work (with recent versions of ember / ember-data).

Second, big thanks to @typeoneerror for a good workaround and for finding out that `[Query Params] cannot be [Computed Properties]`.

Lastly, it's worth noting that using `content` is deprecated and all the references to it should be updated.

**Comment by typeoneerror** [comment]
@sdhull 👍

I'm still on Ember 2.4 so this is working fine for me now, but I plan on migrating over to https://github.com/offirgolan/ember-light-table when I get some time to upgrade Ember.

**Comment by sdhull** [comment]
Another problem I've found is that `page` isn't reset when you navigate away and then navigate back (because controllers aren't necessarily torn down when you navigate away).

**Comment by broerse** [comment]
@sdhull I upgraded to 2.17 and it stopped working. Will try to take what changed.
